### ISSUE TITLE
Circular dependencies in class declarations leads to Unknown type argument (regression)

### ISSUE THREAD

**Issue by relsunkaev** [microsoft/pyright#3566]
**Describe the bug**
As of version 1.1.253, pyright can't detect annotations for certain parts of SQLAlchemy when using [sqlalchemy-stubs](https://github.com/dropbox/sqlalchemy-stubs) or [sqlalchemy2-stubs](https://github.com/sqlalchemy/sqlalchemy2-stubs). This error is not present in 1.1.252.

**To Reproduce**
Install `SQLAlchemy==1.4.37` along with `sqlalchemy2-stubs==0.0.2a23`. The following code snippet
```python
from sqlalchemy.orm import Mapped

class Table:
    id: Mapped[int]

Table.id.asc()
```
will yield the following error
```
/.../test.py
  /.../test.py:6:1 - error: Type of "asc" is partially unknown
    Type of "asc" is "() -> Unknown" (reportUnknownMemberType)
```

**Expected Behavior**
Type check passes.

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the bug as reported, and I've narrowed it down to [this commit](https://github.com/microsoft/pyright/commit/bc4fa7e22bad56b019d2317adc17eebf2885389d). More investigation is needed.

**Comment by erictraut** [comment]
The problem is related to a circular dependency within the sqlalchemy type stubs. The class `ColumnElement` derives from `ColumnOperators[_TE]`, but the class `ColumnOperators` derives from `Operators[ColumnElement[_TE]]`.

This type of circularity is something that a type checker should be able to handle. Clearly, there are other forms like `class A(B)` and `class B(A)` that should not be allowed even if they are declared in a type stub (which supports forward references). Differentiating between "allowable" and "not allowable" circularity will be key to fixing this issue.

**Comment by erictraut** [comment]
I've backed out the change that caused this regression and solved that problem in a different manner. I've confirmed that this issue no longer exists and that pyright now correctly handles this flavor of circular dependency between classes.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.254, which I just published. It will also be included in the next release of pylance.
